xpti/include/xpti_trace_framework.hpp

```diff
@@ -137,73 +137,34 @@ class PlatformHelper {
     return error;
   }
 
+  std::string getEnvironmentVariable(const std::string &var) {
 #if defined(_WIN32) || defined(_WIN64)
-  std::wstring getEnvironmentVariable(const std::string &var) {
-    wchar_t str[WIN_PATH_MAX];
-    auto wvar = new wchar_t[var.size() + 1];
-    size_t nchars, size = var.size() + 1;
-    // Prepare the environment variable name that is in string format to be wide
-    // char format so it can be used with the API that takes in wide char and
-    // returns wide char
-    auto failed = mbstowcs_s(&nchars, wvar, size, var.c_str(), size - 1);
-    // The value of the environment variable will be in 'str'
-    DWORD bytes_returned = GetEnvironmentVariableW(wvar, str, WIN_PATH_MAX);
-    delete[] wvar;
-    if ((bytes_returned > 0 && bytes_returned < WIN_PATH_MAX) && !failed) {
-      return str;
-    } else {
-      // If error message needs to be printed,
-      // std::string error_str = getLastError();
-      // can be used.
-      return {};
-    }
-  }
-
-#else // Linux and MacOSX
-  std::wstring getEnvironmentVariable(const std::string &var) {
-    const char *val = std::getenv(var.c_str());
-    if (val == 0) {
-      return {};
-    } else {
-      const size_t size = strlen(val) + 1;
-      wchar_t *buff = new wchar_t[size];
-      size_t conv_code = mbstowcs(buff, val, size);
-      if (conv_code > 0) {
-        // Conversion succeeded and the
-        // the size of the string is valid
-        std::wstring conv_string = buff;
-        delete[] buff;
-        return conv_string;
-      } else {
-        // We got an error code of -1
-        // or a NULL string?
-        delete[] buff;
+    size_t requiredSize;
+    // Get the size required to hold the value of the environment variable
+    // defined by 'var'
+    getenv_s(&requiredSize, NULL, 0, var.c_str());
+    if(requiredSize == 0) {
+        // The variable we seek is not defined
         return {};
-      }
     }
-  }
-
-  std::string convWChar2MBChar(std::wstring &w) {
-    char buffer[PATH_MAX];
-    int ret;
-
-    int max_width = (w.size() < PATH_MAX) ? w.size() : PATH_MAX;
-    ret = wcstombs(buffer, &w[0], max_width);
-    // If size is less than PATH_MAX, it is
-    // NULL terminated, else we termintate it
-    // and the file path will be truncated.
-    if (ret == PATH_MAX) {
-      buffer[PATH_MAX - 1] = '\0';
-    } else {
-      // However, the string is not
-      // NULL terminated, so we do
-      // it anyway
-      buffer[ret] = '\0';
+    // If the variable exists, then get the value
+    char *value = new char[requiredSize+1];
+    std::string retValue;
+    if (value) {
+      if (!getenv_s(&requiredSize, value, requiredSize, var.c_str()))
+        retValue = value;
+      delete [] value;
     }
 
-    return buffer;
-  }
+    return retValue;;
+#else
+    const char *val = std::getenv(var.c_str());
+    if (val)
+      return val;
+    else
+      return {};
 #endif
+  }
```

Comment:
Congratulation for the simplification.
But at the end, it is not zen enough.
What about just whatever the OS:
```c++
std::string getEnvironmentVariable(const std::string &var) {
  const char* env_p = std::getenv(var.c_str());
  // Return the variable value if it does exist, otherwise the empty string
  return env_p ? env_p : {};
}
```
Is there something lacking?

Reply:
After thinking to it last night I am unsure my solution above does compile.
After checking, it does not because there is no type unification through `?:`

https://godbolt.org/z/-4Ttpj

So either
```c++
std::string getEnvironmentVariable(const std::string &var) {
  const char* env_p = std::getenv(var.c_str());
  // Return the variable value if it does exist, otherwise the empty string
  return env_p ? env_p : std::string {};
}
```
or
```c++

std::string getEnvironmentVariable_2(const std::string &var) {
  const char* env_p = std::getenv(var.c_str());
  // Return the variable value if it does exist, otherwise the empty string
  if (env_p)
    return  env_p;
  return {};
}
```

With pattern matching (in C++23?) the code will be even shorter... :-)

Reply:
After thinking to it last night I am unsure my solution above does compile.
After checking, it does not because there is no type unification through `?:`

https://godbolt.org/z/-4Ttpj

So either
```c++
std::string getEnvironmentVariable(const std::string &var) {
  const char* env_p = std::getenv(var.c_str());
  // Return the variable value if it does exist, otherwise the empty string
  return env_p ? env_p : std::string {};
}
```
or
```c++

std::string getEnvironmentVariable_2(const std::string &var) {
  const char* env_p = std::getenv(var.c_str());
  // Return the variable value if it does exist, otherwise the empty string
  if (env_p)
    return  env_p;
  return {};
}
```

With pattern matching (in C++23?) the code will be even shorter... :-)

Reply:
After thinking to it last night I am unsure my solution above does compile.
After checking, it does not because there is no type unification through `?:`

https://godbolt.org/z/-4Ttpj

So either
```c++
std::string getEnvironmentVariable(const std::string &var) {
  const char* env_p = std::getenv(var.c_str());
  // Return the variable value if it does exist, otherwise the empty string
  return env_p ? env_p : std::string {};
}
```
or
```c++

std::string getEnvironmentVariable_2(const std::string &var) {
  const char* env_p = std::getenv(var.c_str());
  // Return the variable value if it does exist, otherwise the empty string
  if (env_p)
    return  env_p;
  return {};
}
```

With pattern matching (in C++23?) the code will be even shorter... :-)

Reply:
> Congratulation for the simplification.
> But at the end, it is not zen enough.
> What about just whatever the OS:
> 
> ```c++
> std::string getEnvironmentVariable(const std::string &var) {
>   const char* env_p = std::getenv(var.c_str());
>   // Return the variable value if it does exist, otherwise the empty string
>   return env_p ? env_p : {};
> }
> ```
> 
> Is there something lacking?

The windows documentation suggests that the secure version is getenv_s, hence the OS specific use here. I could not find documentation on whether std::getenv is secure on Linux, so perhaps the same "secure" code could be used for both OSes.

Documentation on [getenv](https://en.cppreference.com/w/c/program/getenv) and issues 
discussed in [Stackoverflow]( https://stackoverflow.com/questions/48568707/getenv-this-function-or-variable-may-be-unsafe-really) seem to indicate that both unsecure and secure versions have issues.

Reply:
Done! Picked the first approach, but please see the stackoverflow discussion linked earlier.

Reply:
Sorry for the spam above. GitHub was down this morning but it seems to have remember some of the times I hit the `Comment` button... :-/

Reply:
@keryell No worries, I only read one of them :-)  Thank you for the comprehensive review!

Reply:
Thank you for the links to StackOverflow.
Conclusion: you should not trust the old MS C API and just look at the C++11 `std::getenv()` instead. \o/
https://en.cppreference.com/w/cpp/utility/program/getenv : "This function is thread-safe" [in the case of concurrent reads].
Amazing that the old MS C API implementation used to be that bad...

Reply:
@keryell I believe Klockwork flags std::getenv as unsafe on Windows, so I'd prefer to leave it it stands currently with the use of the safe version for Windows.

Reply:
> @keryell I believe Klockwork flags std::getenv as unsafe on Windows, so I'd prefer to leave it it stands currently with the use of the safe version for Windows.

What do you mean by "I believe"?

Can you add some facts about it?
Which version of Klockwork flagging which version of the Windows C++ compiler and library? SYCL is about very modern C++ (not even talking about C) while Perforce seems more at targeting legacy. If at the end you are adding some complex code for a problem which disappeared 10 years ago and that the workaround introduces some subtle bugs and prevent up-streaming, it might worth looking twice. :-)

Reply:
Klockwork 10.3 reported issues against std::getenv when compiled with VS2015. I recall @romanovvlad  mentioning that they target C++11 compliant compilers, which would be VS2015 on Windows. Using getenv_s currently avoids getting a hit.

Reply:
Let's summarize... Klockwork 10.3, a 2015 version of this tool, mentioned a problem with an obsolete version of MS C++ compiler/library claiming supporting C++11 but not really? Any compliant C++11 implementation should not gave this bug. @erichkeane and @rolandschulz : can you clarify this?

Reply:
> Let's summarize... Klockwork 10.3, a 2015 version of this tool, mentioned a problem with an obsolete version of MS C++ compiler/library claiming supporting C++11 but not really? Any compliant C++11 implementation should not gave this bug. @erichkeane and @rolandschulz : can you clarify this?

Since I do not have the Klockwork reports from a couple of months ago and my recollection of the error I remember seeing being fuzzy, I decided to write a small [example](https://godbolt.org/z/iBP_WU) and test it with VS 2015 compiler.

Reply:
Thanks for experimenting. Interestingly there is no error with https://godbolt.org/z/xgQ5wK
But anyway, rather than randomly shooting in the dark, ask your local colleagues above who are in the ISO C++ committee about. In the worse case they can ask Microsoft ISO C++ committee members what is the right compiler & option supporting a modern standard.

Reply:
If you enable the -sdl option which will enable additional security checks, it will give you the error.

Reply:
Great, then you have found the solution: do not use the `/sdl` option! :-)
After digging into this for a few hours, I think we can reach some consensus.
The MSVC C++ library is now open-sourced but the implementation just wrap the C implementation:
https://github.com/microsoft/STL/blob/master/stl/inc/cstdlib#L13
https://github.com/microsoft/STL/blob/master/stl/inc/cstdlib#L69
so no implementation detail there... :-(
But note that the warning in your example says:
> <source>(19): error C4996: 'getenv': This function or variable may be unsafe. Consider using _dupenv_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
So it does not say that it is unsafe for sure in this context.
By the way, there are some creative comments about the interpretation of "deprecated" in 
- https://docs.microsoft.com/en-us/cpp/c-runtime-library/security-features-in-the-crt?view=vs-2019
> In this context, "deprecated" just means that a function's use is not recommended; it does not indicate that the function is scheduled to be removed from the CRT.
- https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4996?view=vs-2019
> Important
> This warning is always a deliberate message from the author of the header file that declares the symbol. Don't use the deprecated symbol without understanding the consequences.
Some C runtime library functions are deprecated because they're insecure and have a more secure variant. 

The latest ISO C++ standard draft says
> 17.13 Other runtime support [support.runtime]
> 2 Calls to the function getenv (17.2.2) shall not introduce a data race (16.5.5.10) provided that nothing modifies the environment. [Note: Calls to the POSIX functions setenv and putenv modify the environment.
— end note]

I guess you are not planning to implement this race condition on purpose in this runtime...
If someone is doing it under a debugger, well... anything in the program can fall apart anyway... :-)

But it is possible to remove the warning if you are serious about using `/sdl` option, like in https://godbolt.org/z/AvoyP6
```
#include <cstdlib>
#include <iostream>
#include <string>

/// Insert something when compiled with msvc
/// https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros
#ifdef _MSC_VER
#define __SYCL_INSERT_IF_MSVC(x) x
#else
#define __SYCL_INSERT_IF_MSVC(x)
#endif

int main() {
  // Remove irrelevant deprecation warning in this context
  // https://docs.microsoft.com/en-us/cpp/c-runtime-library/security-features-in-the-crt
  // https://docs.microsoft.com/en-us/cpp/build/reference/sdl-enable-additional-security-checks
  __SYCL_INSERT_IF_MSVC(__pragma(warning(suppress : 4996)))
  const auto bar = std::getenv("MY_VARIABLE");
  std::string env { bar ? bar : "" };
  std::cout << "env variable MY_VARIABLE = " << env << "\n";
}
```
and keep the comments and improve them to explain how we ended up into this source code...

Reply:
@keryell  Unfortunately, /sdl is now required in most companies and disabling a warning so it doesn't cause a failure is something  I am uncomfortable with.  @bader, @alexbatashev, @romanovvlad  - I'll make this change, but I will leave the other code in there commented.   In case you are against suppressing the warning, we can always switch back to the other code.

Reply:
I am still discussing with Microsoft C & C++ runtime implementors on this issue.

Reply:
@keryell Please keep me posted and I will ensure the conversation is reflected in the code. Thanks!